src/python/pants/engine/internals/uuid.py

```diff
@@ -0,0 +1,26 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import uuid
+from dataclasses import dataclass
+
+from pants.engine.rules import RootRule, uncacheable_rule
+
+
+@dataclass(frozen=True)
+class UUIDRequest:
+    pass
+
+
+@uncacheable_rule
+async def generate_uuid(_: UUIDRequest) -> uuid.UUID:
+    """A rule to generate a UUID.
+
```

Comment:
So, technically "uncacheable" is "once per session/run" right now. If two requests arrived with the same `UUIDRequest` within the same run, they'd get the same UUID out. That's fine for this use case, but possibly worth describing here: if it mattered to a consumer you might want to give the `UUIDRequest` a seed parameter.

Reply:
Maybe just renaming this to "RunID" (and returning the actual runid we use for reporting, if we can get hold of it...)?

Reply:
`RunId` sounds good. We pass it to the engine here:

https://github.com/pantsbuild/pants/blob/669b5f3845944fb80c705f1537d1bc6d6bce8dd1/src/python/pants/bin/local_pants_runner.py#L89-L90

I imagine you would need to create a singleton rule to expose this, similar to this:

https://github.com/pantsbuild/pants/blob/669b5f3845944fb80c705f1537d1bc6d6bce8dd1/src/python/pants/init/engine_initializer.py#L274-L276

Reply:
OK the RunID thing is more complicated, as those appear to now be generated in Rust. So we probably want an intrinsic to expose it. That could be useful at some point, but is distinct from this, so I'll just comment it for now.

Reply:
Actually this was pretty easy to fix after all, to have the expected behavior, so I've done so. PTAL

Reply:
Neat.

Reply:
Plus a test, for good measure.